### ISSUE TITLE
Clang-Tidy: misc-redundant-expression false positive with operator== overload

### ISSUE THREAD

**Issue by pavelkryukov** [llvm/llvm-project#54011]
For the following example:

```c++
enum Enum {
    A = 0,
    B = 1,
    C,
    D
};

struct Struct {
    Enum value;
    bool operator==( Enum x) const { return value == x; }
};

Struct foo( const Struct& value)
{
    if ( value == A || value == B)
        return value;

    return Struct{ D};
}
```

a wrong warning message is generated: https://godbolt.org/z/br8bsTaPo

```
<source>:15:21: warning: equivalent expression on both sides of logical operator [misc-redundant-expression]
    if ( value == A || value == B)
                    ^
1 warning generated.
```

**Comment by FabianWolff** [comment]
This false positive appears even without the `enum`:
```c++
struct Struct {
    int value;
    bool operator==(int x) const { return value == x; }
};

Struct foo( const Struct& value)
{
    if (value == 42 || value == 43)
        return value;

    return Struct{ 0 };
}
```
```
<source>:8:21: warning: equivalent expression on both sides of logical operator [misc-redundant-expression]
    if (value == 42 || value == 43)
                    ^
1 warning generated.
```
Fixed by [D122111](https://reviews.llvm.org/D122111).

**Comment by FabianWolff** [comment]
Fixed in https://github.com/llvm/llvm-project/commit/07675b0b38e930c2c582122ac93efcdf7859a772.

**Comment by pavelkryukov** [comment]
Thanks @FabianWolff!
Years ago I've submitted a bug #35853. Could you please confirm it is fixed with this patch?
